### PR TITLE
Move executeCode out of routeLoaders, into shared space in Stream

### PR DESCRIPTION
**Pull request by franknoirot**
Resolves re-execution of KCL code when opening and closing the project settings, and potentially more racing behavior.

## Reasoning

We need to re-execute the KCL code when navigating between projects and files, zoom-to-fit on the resulting geometry, and unpause the stream's video element. Until now, we have kept the `executeCode` and zoom-to-fit portions of that process in [the route loader](https://reactrouter.com/en/main/route/loader) for the `/file/**` route, as an unawaited promise ([here](https://github.com/KittyCAD/modeling-app/blob/main/src/lib/routeLoaders.ts#L109)).

We relied on this because route loaders run before rendering, so it seemed a good place to put execution. This is problematic because route loaders run in parallel and can complete at varying times, and by not awaiting the promise we introduced the possibility that the app could be rendering and acting as if execution completed while it was still pending.

Furthermore, this route loader re-fires every time a route matching its pattern is hit, so it re-fired every time the user opened their settings within a project, since that lives at `/file/<your-files-path>/settings`, causing an expensive re-execution and re-render of the stream.

By co-locating the code execution with the stream start, we move closer to a common code path for all code re-execution.

## Future work

I do not love having this code live so far down the component tree, so coupled to the stream. If there were a way for the Stream to know for certain that we have executed the code somewhere higher up—such as in [FileMachineProvider](https://github.com/KittyCAD/modeling-app/blob/franknoirot/no-execute-in-loaders/src/Router.tsx#L98)—then we could let the Stream only worry about the unpausing of the video element it is bound to, and not the KCL execution and zoom-to-fit call site which seems more general.

## Demo

https://github.com/user-attachments/assets/e88bdf34-d530-49c4-97ec-559567511804

### CODE REVIEW

File: src/components/Stream.tsx [hunk 1]
```diff
@@ -15,6 +15,10 @@ import {
   EngineConnectionStateType,
   DisconnectingType,
 } from 'lang/std/engineConnection'
+import { useFileContext } from 'hooks/useFileContext'
```
Comment (jtran) [inline src/components/Stream.tsx:18]:
```suggestion
```
GitHub is warning that this line is never used.

File: src/lib/routeLoaders.ts [hunk 2]
```diff
@@ -105,9 +105,6 @@ export const fileLoader: LoaderFunction = async ({
     codeManager.updateCurrentFilePath(current_file_path)
     codeManager.updateCodeStateEditor(code)
 
-    // We don't want to call await on execute code since we don't want to block the UI
-    kclManager.executeCode(true)
```
Comment (jtran) [inline src/lib/routeLoaders.ts:109]: I like this change. Not `await`ing seems like asking for race conditions.

Can you remove the `kclManager` import that seems no longer used?

File: src/components/Stream.tsx [hunk 3]
```diff
@@ -32,13 +36,47 @@ export const Stream = () => {
...
+   * Subscribe to execute code when the file changes
+   * but only if the scene is already ready.
+   * See onSceneReady for the initial scene setup.
+   */
+  useEffect(() => {
+    if (engineCommandManager.engineConnection?.isReady() && file?.path) {
+      console.log('execute on file change')
+      executeCodeAndPlayStream()
+    }
+  }, [file?.path])
```
Comment (jtran) [inline src/components/Stream.tsx:78]: Does this need to also depend on the `engineConnection`'s readiness?

Comment (lf94) [reply src/components/Stream.tsx:78]: Yeah I think adding `engineCommandManager.engineConnection` to the dependencies list would be a good idea here.

**Review by lf94 — APPROVED** [review]
Looks good! (Other than what Jon commented)

**Review by jtran — APPROVED** [review]

**Comment by franknoirot** [comment]
I think I can't have that added dependency in that useEffect. I'll check in the morning

**Review by Irev-Dev — APPROVED** [review]

**Pull request merged**